sdk/core/System.ClientModel/src/Pipeline/ClientLoggingPolicy.cs

```diff
@@ -0,0 +1,455 @@
+    {
+        Singleton.Request(request, requestId ?? string.Empty, _assemblyName, Sanitizer!);
+    }
+
+    private async Task LogRequestContent(PipelineRequest request, string? requestId, bool async, CancellationToken cancellationToken)
+    {
+        // If logging content is disabled, or the request content is null, or informational logs are not enabled, return
+        // Checking the log level prevents expensive operations from being executed
+        if (!_logContent || request.Content == null || !Singleton.IsEnabled(EventLevel.Informational, EventKeywords.All))
+        {
+            return; // nothing to log
+        }
+
+        // Convert binary content to bytes
+        using var memoryStream = new MaxLengthStream(_maxLength);
+        if (async)
+        {
+            await request.Content.WriteToAsync(memoryStream, cancellationToken).ConfigureAwait(false);
+        }
+        else
+        {
+            request.Content.WriteTo(memoryStream, cancellationToken);
+        }
+        var bytes = memoryStream.ToArray();
+
+        // Try to extract a text encoding from the headers
+        Encoding? requestTextEncoding = null;
+
+        if (request.Headers.TryGetValue("Content-Type", out var contentType) && contentType != null)
+        {
+            ContentTypeUtilities.TryGetTextEncoding(contentType, out requestTextEncoding);
+        }
+
+        // Log to event source
+        Singleton.RequestContent(requestId ?? string.Empty, bytes, requestTextEncoding);
+    }
+
+    private void LogResponse(PipelineResponse response, string? responseId, double elapsed)
+    {
+        bool isError = response.IsError;
+        if (isError)
+        {
+            Singleton.ErrorResponse(response, responseId ?? string.Empty, Sanitizer!, elapsed);
+        }
+        else
+        {
+            Singleton.Response(response, responseId ?? string.Empty, Sanitizer!, elapsed);
+        }
+    }
+
+    private async Task LogResponseContent(PipelineResponse response, string? responseId, bool bufferResponse, bool async, CancellationToken cancellationToken)
+    {
+        // If logging content is disabled, or the request content is null, or informational logs are not enabled, return.
+        // Checking the log level prevents expensive operations from being executed
+        // TODO - consider moving this down to LoggingStream
+        var logErrorResponseContent = response.IsError && Singleton.IsEnabled(EventLevel.Warning, EventKeywords.All);
+        var logResponseContent = Singleton.IsEnabled(EventLevel.Informational, EventKeywords.All) || logErrorResponseContent;
+
+        if (!_logContent || !logResponseContent)
+        {
+            return; // nothing to log
+        }
+
+        // Determine if the content stream should be wrapped in a logging stream, which logs the stream in blocks as it is being read
+        var wrapContent =
+            !bufferResponse // Content is not buffered - content should be in the content stream
```

Comment:
Would it be ok to rename this variable to something like `contentBuffered` ?  That feels a little easier to read/follow to me, rather than needing to remember that `message.BufferResponse` is the name of the property that tells the transport whether or not to buffer the response.  Then, this statement becomes something like:

```csharp
bool wrapContent = !contentBuffered && !response.ContentStream.CanSeek;
```

Or, you could do it the other way:

```csharp
bool useBufferedContent = contentBuffered || response.ContentStream.CanSeek;
```

Or, since we're allocating a new stream either way, maybe we just remove the CanSeek check and always wrap it in a logging stream if the content isn't buffered?  Is there a downside to doing that?  The content doesn't get logged until the app reads the content stream, but is that bad?

Alternatively (sorry, nerding a bit here 🙂), we could check these things sequentially and exit the routine after each check, e.g.:

1. if content buffered, log and return;
2. if stream.CanSeek, log and return;
3. wrap the content stream and return;

What do you find most readable/clearest?  I just struggle with wrapping my head around multi-level if/else logic, so I like finding the cleanest way -- it usually helps others coming to the code for the first time or needing to maintain it without a lot of ramp-up time in the future.

Reply:
I updated this to checking things sequentially. I think it looks much nicer this way! Thanks for the suggestion!